nemo/collections/asr/models/wav2vec/wav2vec_model.py

```diff
@@ -0,0 +1,541 @@
+            features = self.post_extract_proj(features)
+
+        features = self.dropout_input(features)
+        unmasked_features = self.dropout_features(unmasked_features)
+
+        if self.input_quantizer:
+            features, prob_ppl, cur_codebook_temp = self.input_quantizer(features)
+            features = self.project_inp(features)
+        if mask:
+            x, mask_indices = self.apply_mask(features, padding_mask)
+            if mask_indices is not None:
+                y = unmasked_features[mask_indices].view(unmasked_features.size(0), -1, unmasked_features.size(-1))
+            else:
+                y = unmasked_features
+        else:
+            x = features
+            y = unmasked_features
+            mask_indices = None
+
+        x = self.encoder(x, padding_mask=padding_mask)
+
+        if features_only:
+            return x, padding_mask
+
+        if self.quantizer:
+            y, prob_ppl, cur_codebook_temp = self.quantizer(y)
+            y = self.project_q(y)
+
+            if self.negatives_from_everywhere:
+                neg_cands, *_ = self.quantizer(unmasked_features)
+                sampled_negatives, _ = self.sample_negatives(neg_cands, y.size(1))
+                sampled_negatives = self.project_q(sampled_negatives)
+            else:
+                sampled_negatives, _ = self.sample_negatives(y, y.size(1))
+
+            if self.codebook_negatives > 0:
+                cb_negs = self.quantizer.sample_from_codebook(y.size(0) * y.size(1), self.codebook_negatives)
+                cb_negs = cb_negs.view(self.codebook_negatives, y.size(0), y.size(1), -1)  # order doesnt matter
+                cb_negs = self.project_q(cb_negs)
+                sampled_negatives = torch.cat([sampled_negatives, cb_negs], dim=0)
+        else:
+            y = self.project_q(y)
+
+            if self.negatives_from_everywhere:
+                sampled_negatives, _ = self.sample_negatives(unmasked_features, y.size(1))
+                sampled_negatives = self.project_q(sampled_negatives)
+            else:
+                sampled_negatives, _ = self.sample_negatives(y, y.size(1))
+
+        x = x[mask_indices].view(x.size(0), -1, x.size(-1))
+
+        if self.target_glu:
+            y = self.target_glu(y)
+            sampled_negatives = self.target_glu(sampled_negatives)
+
+        x = self.final_proj(x)
+        output = Wav2VecEncoderOutput(
+            logits=x,
+            targets=y,
+            sampled_negatives=sampled_negatives,
+            padding_mask=padding_mask,
+            features_penalty=features_penalty,
+            probs_ppl=prob_ppl,
+            cur_codebook_temp=cur_codebook_temp,
+        )
+        return output
+
+    def extract_features(self, source, audio_lengths, mask=False):
+        padding_mask = self._create_padding_mask(audio_lengths)
+        return self.forward(source, padding_mask, mask=mask, features_only=True)
+
+    def remove_pretraining_modules(self):
```

Comment:
How about instead of removing those modules, we set a variable to disable them and not erase them. This erasing may cause issues later as we may want to store the final model which do not contain these modules anymore but may be needed to continue the pretraining. When we delete the modules, they are gone.

Reply:
Great point :)